### PR TITLE
Feature/178

### PR DESCRIPTION
**Pull request by NicholasMcCarthy**
ConvE implemented. 

- Adds new model ConvE. 
- Adds OneToNAdapter (required for ConvE)
- Adds BCELoss (binary cross entropy, required for ConvE)
- Results table includes comparison of ConvE with literature reported results as well as standard evaluation.

### CODE REVIEW

File: docs/experiments.rst [hunk 1]
```diff
@@ -91,14 +91,57 @@ FB15K-237
...
+                                                         conv_kernel_size: 3,
+                                                         dropout_embed: 0.2,
+                                                         dropout_conv: 0.1,
+                                                         dropout_dense: 0.3,
+                                                         use_batchnorm: True,
+                                                         use_bias: True};
+                                                         seed: 0;
+                                                         batches_count: 100;
+
+  ConvE*     234      0.32   0.23     0.35     0.50       k: 200;
```
Comment (sumitpai) [inline docs/experiments.rst]: Should we call ConvE*->ConvE and vice versa. Since the current conve* is the original paper implementation?

File: ampligraph/latent_features/models/ConvE.py [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+    """
+
+    def __init__(self,
+                 k=constants.DEFAULT_EMBEDDING_SIZE,
+                 eta=constants.DEFAULT_ETA,
+                 epochs=constants.DEFAULT_EPOCH,
+                 batches_count=constants.DEFAULT_BATCH_COUNT,
+                 seed=constants.DEFAULT_SEED,
+                 embedding_model_params={'conv_filters': 32,
+                                         'conv_kernel_size': 3,
```
Comment (sumitpai) [inline ampligraph/latent_features/models/ConvE.py]: use constants.DEFAULTS

File: ampligraph/latent_features/models/ConvE.py [hunk 3]
```diff
@@ -0,0 +1,1158 @@
...
+                self.bias = tf.get_variable('activation_bias', shape=[1, len(self.ent_to_idx)],
+                                            initializer=tf.zeros_initializer(), dtype=tf.float32)
+
+        else:
+            raise NotImplementedError('ConvE not implemented when dealing with large graphs.')
+
+    def _get_model_loss(self, dataset_iterator):
+        """Get the current loss including loss due to regularization.
+        This function must be overridden if the model uses combination of different losses(eg: VAE).
+        This function must be overridden if the model uses combination of different losses(eg: VAE).
```
Comment (sumitpai) [inline ampligraph/latent_features/models/ConvE.py]: repeated twice

File: ampligraph/latent_features/models/ConvE.py [hunk 4]
```diff
@@ -0,0 +1,1158 @@
...
+            # look up embeddings from input training triples
+            e_s_pos, e_p_pos, e_o_pos = self._lookup_embeddings(self.x_pos_tf)
+
+            # Get positive predictions
+            self.y_pred = self._fn(e_s_pos, e_p_pos, e_o_pos)
+
+            # Label smoothing and/or weighting is applied within Loss class
+            loss = self.loss.apply(self.y_true, self.y_pred)
+
+            if self.regularizer is not None:
```
Comment (sumitpai) [inline ampligraph/latent_features/models/ConvE.py]: Should we also regularize the weights?

File: ampligraph/datasets/oneton_adapter.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+            indicates which dataset to use
+        use_filter : bool
+            Flag to indicate whether to return the one-hot outputs are generated from filtered or unfiltered datasets
+
+        Returns
+        -------
+
+        test_triples : nd-array, shape=[batch_size, 3]
+            A set of triples from the dataset type specified, that include the predicate currently returned in batch.
+        batch_triples : nd-array of shape (N, 3), where N is number of unique entities.
```
Comment (sumitpai) [inline ampligraph/datasets/oneton_adapter.py]: fix the doc string (return sizes)

File: ampligraph/latent_features/models/ConvE.py [hunk 6]
```diff
@@ -0,0 +1,1158 @@
...
+
+                        if len(X_test) == 0:
+                            # If X_test is empty, reset accumulated scores and continue
+                            scores_matrix_accum, scores_filter_accum = [], []
+                            continue
+
+                        scores_matrix = np.concatenate(scores_matrix_accum)
+                        scores_filter = np.concatenate(scores_filter_accum)
+
+                        for i, x in enumerate(X_test):
```
Comment (sumitpai) [inline ampligraph/latent_features/models/ConvE.py]: there is still an issue. While the mrr will not get affected here (since we average the ranks), it would still affect the ranks of individual triples (consider the case : test set made of [<s, p, 7>, <s, p, 3>]

File: ampligraph/evaluation/protocol.py [hunk 7]
```diff
@@ -493,6 +494,7 @@ def evaluate_performance(X, model, filter_triples=None, verbose=False, strict=Tr
 
         - 's': corrupt only subject.
         - 'o': corrupt only object.
+        - '1-N': object only corruption using the fast strategy given in `cite`Dettmers2016. Only valid for ConvE model.
```
[the remark this answers is not in the document]
Comment (NicholasMcCarthy) [reply ampligraph/evaluation/protocol.py]: Done.

[the remark this answers is not in the document]
Comment (NicholasMcCarthy) [reply ampligraph/evaluation/protocol.py]: Removed '1-N' as an option since it's just a different implementation, not a different evaluation.

File: ampligraph/evaluation/protocol.py [hunk 8]
```diff
@@ -541,16 +543,25 @@ def evaluate_performance(X, model, filter_triples=None, verbose=False, strict=Tr
...
     dataset_handle = None
     # try-except block is mainly to handle clean up in case of exception or manual stop in jupyter notebook
     try:
         logger.debug('Evaluating the performance of the embedding model.')
         if isinstance(X, np.ndarray):
+
             X_test = filter_unseen_entities(X, model, verbose=verbose, strict=strict)
 
-            dataset_handle = NumpyDatasetAdapter()
+            if isinstance(model, ConvE):
```
[the remark this answers is not in the document]
Comment (NicholasMcCarthy) [reply ampligraph/evaluation/protocol.py]: All options except 's+o' are valid for ConvE. An exception is raised in ConvE if eval_protocol is 's+o'

File: ampligraph/evaluation/protocol.py [hunk 9]
```diff
@@ -883,7 +894,7 @@ def select_best_model_ranking(model_class, X_train, X_valid, X_test, param_grid,
                               early_stopping_params=None, use_test_for_selection=False, entities_subset=None,
                               corrupt_side='s+o', use_default_protocol=True, retrain_best_model=False, verbose=False):
     """Model selection routine for embedding models via either grid search or random search.
-    
+
```
[the remark this answers is not in the document]
Comment (NicholasMcCarthy) [reply ampligraph/evaluation/protocol.py]: It's removed as an option instead.

File: ampligraph/latent_features/models/ConvE.py (hunk 2 above)
Comment (NicholasMcCarthy) [reply ampligraph/latent_features/models/ConvE.py]: Done.

File: ampligraph/evaluation/protocol.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (NicholasMcCarthy) [reply ampligraph/evaluation/protocol.py]: Is the issue here the 1-N corruption details differences, or the use_default_protocol issue? I've resolved any conflicts from that merge, anyway ..

File: ampligraph/latent_features/models/ConvE.py (hunk 4 above)
Comment (NicholasMcCarthy) [reply ampligraph/latent_features/models/ConvE.py]: Good question.

File: ampligraph/datasets/oneton_adapter.py (hunk 5 above)
Comment (NicholasMcCarthy) [reply ampligraph/datasets/oneton_adapter.py]: Docstring details were actually incorrect, so : fixed

File: ampligraph/latent_features/models/ConvE.py (hunk 6 above)
Comment (NicholasMcCarthy) [reply ampligraph/latent_features/models/ConvE.py]: As discussed, not the problem you thought it was .. ;)

File: ampligraph/latent_features/models/ConvE.py (hunk 4 above)
Comment (NicholasMcCarthy) [reply ampligraph/latent_features/models/ConvE.py]: There's no regularization in the ConvE github reference code, so I'll leave it for now.

File: ampligraph/evaluation/protocol.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sumitpai) [reply ampligraph/evaluation/protocol.py]: That's fine. I thought there was a new mode corrupt_side='1-N' as it was in the docstring. But I guess you have changed it to use one of the existing modes. So it doesn't hold.
